### PR TITLE
rework test to use libhostspriv.so

### PR DESCRIPTION
**Pull request by bobvanderlinden**
Currently `hosts.c` includes tests using the `COMPILE_FOR_TEST` define.
It only tests for `lookup_alias`, but not the `LD_PRELOAD` hooking.

This PR attempts to move the test code to a separate executable, which is used with `LD_PRELOAD` and `test_hosts` to test the full flow of libhostspriv.

Currently it only tests for `getaddrinfo`.

### CODE REVIEW

File: test_getaddrinfo.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
+	hints.ai_socktype = SOCK_DGRAM; /* Datagram socket */
+	hints.ai_flags = 0;
+	hints.ai_protocol = 0;          /* Any protocol */
+
+	int s = getaddrinfo(hostname, NULL, &hints, &result);
+	if (s != 0) {
+		out[0] = '\0';
```
Comment (figiel) [inline test_getaddrinfo.c]: maybe just initialize result in assert_resolve.

File: hosts.c [hunk 2]
```diff
@@ -134,84 +122,4 @@ FUN_BEGIN(gethostbyname2_r, int, name, char *name, int a, void *b, void *c, size
...
-
-	return 0;
-}
-
-int main() {
-	return test_hosts();
-}
-
-#endif /* COMPILE_FOR_TEST */
+FUN_END(name, a);
```
Comment (figiel) [inline hosts.c]: nit: please keep the newline at the end of file.

File: test_getaddrinfo.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		return;
+	}
+	if (result == NULL) {
+		out[0] = '\0';
+		return;
+	}
+
+	struct sockaddr_in *addr;
+	addr = (struct sockaddr_in *)result->ai_addr;
```
Comment (figiel) [inline test_getaddrinfo.c]: Why assume IPv4 here?

File: test_getaddrinfo.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ assert(str1 != NULL); \
+ assert(str2 != NULL); \
+ assert(strcmp(str1, str2)==0); } while (0)
+#define ASSERT assert
+
+void resolve(const char *hostname, char *out) {
+	struct addrinfo hints;
+	struct addrinfo *result;
+
+	memset(&hints, 0, sizeof(hints));
```
Comment (figiel) [inline test_getaddrinfo.c]: I don't think this is needed just initialize hints in declaration (e.g. hints = {}).

File: test_getaddrinfo.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
+	hints.ai_socktype = SOCK_DGRAM; /* Datagram socket */
+	hints.ai_flags = 0;
+	hints.ai_protocol = 0;          /* Any protocol */
+
+	int s = getaddrinfo(hostname, NULL, &hints, &result);
```
Comment (figiel) [inline test_getaddrinfo.c]: `s` is not used anywhere else and doesn't mean much, maybe just inline getaddrinfo in if.

**Review by figiel — CHANGES_REQUESTED** [review]

File: test_getaddrinfo.c (hunk 3 above)
Comment (bobvanderlinden) [reply test_getaddrinfo.c]: The example I looked up didn't use `inet_ntop`. I now migrated to use `inet_ntop`, which does support ipv6 :+1: Added ipv6 tests as well.

File: hosts.c (hunk 2 above)
Comment (bobvanderlinden) [reply hosts.c]: Fixed this and added .editorconfig to avoid similar future problems.

File: test_getaddrinfo.c (hunk 4 above)
Comment (bobvanderlinden) [reply test_getaddrinfo.c]: Used `{}` to initialize the fields as well, much cleaner indeed.

File: test_getaddrinfo.c (hunk 5 above)
Comment (bobvanderlinden) [reply test_getaddrinfo.c]: Indeed, inlined it now.

File: test_getaddrinfo.c (hunk 1 above)
Comment (bobvanderlinden) [reply test_getaddrinfo.c]: Done :+1:

**Comment by bobvanderlinden** [comment]
I also added a test that 'redirects' hostname resolving `localhost localhostalias`, as I added that in the readme of #6 as well.

**Comment by bobvanderlinden** [comment]
Incorporated your comments. I think it looks much better.

File: test_getaddrinfo.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+#include <assert.h>
+
+#define ASSERT_EQUAL(str1, str2) do { \
+ assert(str1 != NULL); \
+ assert(str2 != NULL); \
+ assert(strcmp(str1, str2)==0); } while (0)
+#define ASSERT assert
+
+const void *getaddr(const struct addrinfo *info) {
```
Comment (figiel) [inline test_getaddrinfo.c]: This likely needs to be static, sorry missed this earlier.

File: test_getaddrinfo.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  if (info->ai_family == AF_INET) {
+    const struct sockaddr_in *addr = (struct sockaddr_in *)info->ai_addr;
+    return &addr->sin_addr;
+  } else if (info->ai_family == AF_INET6) {
```
Comment (figiel) [inline test_getaddrinfo.c]: I'd remove else here, previous if always returns.

**Comment by figiel** [comment]
This looks good, I made some nitpick comments, please fix if you agree and squash the commits.

**Comment by bobvanderlinden** [comment]
Thanks for the comments! I wasn't sure why the getaddr needed to be static, but i added it anyway :+1:

I reworked the commits a bit to have a cleaner history.

**Comment by figiel** [comment]
Static is preferred by default unless you need to access the symbol from another file, otherwise you risk name conflicts.
Thanks for contributions!

**Comment by bobvanderlinden** [comment]
Thank you for the explanation and the merge! :smile::+1:

**Pull request merged**